### PR TITLE
Unify PyTorch mobile's threadpool usage.

### PR DESCRIPTION
**Pull request by AshkanAliabadi**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37243 Unify PyTorch mobile's threadpool usage.**

*** Why ***

As it stands, we have two thread pool solutions concurrently in use in PyTorch mobile: (1) the open source pthreadpool library under third_party, and (2) Caffe2's implementation of pthreadpool under caffe2/utils/threadpool.  Since the primary use-case of the latter has been to act as a drop-in replacement for the third party version so as to enable integration and usage from within NNPACK and QNNPACK, Caffe2's implementation is intentionally written to the exact same interface as the third party version.

The original argument in favor of C2's implementation has been improved performance as a result of using spin locks, as opposed to relinquishing the thread's time slot and putting it to sleep - a less expensive operation up to a point.  That seems to have given C2's implementation the upper hand in performance, hence justifying the added maintenance complexity, until the third party version improved in parallel surpassing the efficiency of C2's implementation as I have verified in benchmarks.  With that advantage gone, there is no reason to continue using C2's implementation in PyTorch mobile either from the perspective of performance or code hygiene.  As a matter of fact, there is considerable performance benefit to be had as a result of using the third party version as it currently stands.

This is a tricky change though, mainly because in order to avoid potential performance regressions, of which I have witnessed none but just in abundance of caution, we have decided to continue using the internal C2's implementation whenever building for Caffe2.  Again, this is mainly to avoid potential performance regressions in production C2 use cases even if doing so results in reduced performance as far as I can tell.

So to summarize, today, and as it currently stands, we are using C2's implementation for (1) NNPACK, (2) PyTorch QNNPACK, and (3) ATen parallel_for on mobile builds, while using the third party version of pthreadpool for XNNPACK as XNNPACK does not provide any build options to link against an external implementation unlike NNPACK and QNNPACK do.

The goal of this PR then, is to unify all usage on mobile to the third party implementation both for improved performance and better code hygiene.  This applies to PyTorch's use of NNPACK, QNNPACK, XNNPACK, and mobile's implementation of ATen parallel_for, all getting routed to the
exact same third party implementation in this PR.

Considering that NNPACK, QNNPACK, and XNNPACK are not mobile specific, these benefits carry over to non-mobile builds of PyTorch (but not Caffe2) as well.  The implementation of ATen parallel_for on non-mobile builds remains unchanged.

*** How ***
    
This is where things get tricky.

A good deal of the build system complexity in this PR arises from our desire to maintain C2's implementation intact for C2's use.

pthreadpool is a C library with no concept of namespaces, which means two copies of the library cannot exist in the same binary or symbol collision will occur violating ODR.  This means that somehow, and based on some condition, we must decide on the choice of a pthreadpool implementation.  In practice, this has become more complicated as a result of all the possible combinations that USE_NNPACK, USE_QNNPACK, USE_PYTORCH_QNNPACK, USE_XNNPACK, USE_SYSTEM_XNNPACK, USE_SYSTEM_PTHREADPOOL and other variables can result in.  Having said that, I have done my best in this PR to surgically cut through this complexity in a way that minimizes the side effects, considering the significance of the performance we are leaving on the table, yet, as a result of this combinatorial explosion explained above I cannot guarantee that every single combination will work as expected on the first try.  I am heavily relying on CI to find any issues as local testing can only go that far.

Having said that, this PR provides a simple non mobile-specific C++ thread pool implementation on top of pthreadpool, namely caffe2::PThreadPool that automatically routes to C2's implementation or the third party version depending on the build configuration.  This simplifies the logic at the cost of pushing the complexity to the build scripts.  From there on, this thread pool is used in aten parallel_for, and NNPACK and family, again, routing all usage of threading to C2 or third party pthreadpool depending on the build configuration.

When it is all said or done, the layering will look like this:

a) aten::parallel_for, uses
b) caffe2::PThreadPool, which uses
c) pthreadpool C API, which delegates to
    c-1) third_party implementation of pthreadpool if that's what the build has requested, and the rabbit hole ends here.
    c-2) C2's implementation of pthreadpool if that's what the build has requested, which itself delegates to
    c-2-1) caffe2::ThreadPool, and the rabbit hole ends here.

NNPACK, and (PyTorch) QNNPACK directly hook into (c). They never go through (b).

Differential Revision: [D21232894](https://our.internmc.facebook.com/intern/diff/D21232894)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -472,7 +472,7 @@ if(USE_PYTORCH_QNNPACK)
 endif()
 
 if(USE_XNNPACK)
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DUSE_XNNPACK -DUSE_INTERNAL_THREADPOOL_IMPL")
```
Comment (AshkanAliabadi) [inline CMakeLists.txt:475]: USE_INTERNAL_THREADPOOL_IMPL means use C2's version from now on.

File: android/pytorch_android/src/main/cpp/pytorch_jni_common.cpp [hunk 2]
```diff
@@ -8,8 +8,7 @@
 
 #include "pytorch_jni_common.h"
 #if defined(__ANDROID__)
-#include <caffe2/utils/threadpool/ThreadPool.h>
-#include <caffe2/utils/threadpool/ThreadPoolMobile.h>
+#include <caffe2/utils/threadpool/pthreadpool-cpp.h>
```
Comment (AshkanAliabadi) [inline android/pytorch_android/src/main/cpp/pytorch_jni_common.cpp:11]: This new threadpool is not mobile specific since we allow NNPACK, QNNPACK, and XNNPACK on non-mobile builds.  In reality this thread pool is just a simple wrapper around pthreadpool so there is nothing mobile specific about it.  I am open to better names if you have any suggestions.

File: aten/src/ATen/ParallelNative.cpp [hunk 3]
```diff
@@ -87,15 +86,15 @@ void _run_with_pool(const std::function<void(int, size_t)>& fn, size_t range) {
...
-  if (pool) {
-    // caffe2::ThreadPool can utilize the current thread.
-    pool->run(fn, range);
-  } else {
-    for (size_t i = 0; i < range; ++i) {
-      fn(0, i);
-    }
-  }
+  caffe2::PThreadPool* const pool = caffe2::pthreadpool();
+  TORCH_INTERNAL_ASSERT(pool, "Invalid thread pool!");
```
Comment (AshkanAliabadi) [inline aten/src/ATen/ParallelNative.cpp:90]: We'll always have an implementation: the third-party version whenever that's available (i.e. platforms other than Windows), and C2's implementation if third-party is not available (which supports Windows.)

File: aten/src/ATen/native/NNPACK.cpp [hunk 4]
```diff
@@ -85,35 +86,7 @@ static bool init_nnpack() {
...
-
-    nnpack_threadpool_ = pthreadpool_create(threads);
-    if (!nnpack_threadpool_) {
-      LOG(WARNING) << "Failed to initialize pthreadpool! Running NNPACK in single-threaded mode.";
-    }
-  }
-
-  return nnpack_threadpool_;
-#endif
+  return caffe2::pthreadpool_();
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/NNPACK.cpp:89]: As mentioned, this will now automatically select between C2's implementation and the third-party depending on build configuration, covering both usages this code replaces.

File: aten/src/ATen/native/quantized/cpu/q_avgpool.cpp [hunk 5]
```diff
@@ -375,7 +375,7 @@ Tensor qnnpack_avg_pool2d(
   CAFFE_ENFORCE(
       setupStatus == pytorch_qnnp_status_success,
       "failed to setup QNNPACK Average Pooling operator");
-  pthreadpool_t threadpool = caffe2::mobile_pthreadpool();
+  pthreadpool_t threadpool = caffe2::pthreadpool_();
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/quantized/cpu/q_avgpool.cpp:378]: The function name ending in underscore is for internal usage - it returns the pthreadpool_t handle of the underlying implementation.  Open to new names if you have suggestions.

File: caffe2/utils/CMakeLists.txt [hunk 6]
```diff
@@ -1,15 +1,8 @@
-# TODO: Add ThreadPoolXNNPACK.cc when XNNPACK integration is updated
-# to pass the actual threadpool ptr instead of nullptr.
 if(INTERN_BUILD_MOBILE AND NOT BUILD_CAFFE2_MOBILE)
-  add_definitions(-DUSE_INTERNAL_THREADPOOL_IMPL)
```
Comment (AshkanAliabadi) [inline caffe2/utils/CMakeLists.txt:4]: Now defaulting to the third party version for mobile, hence not USE_INTERNAL_THREADPOOL_IMPL.

File: caffe2/utils/CMakeLists.txt [hunk 7]
```diff
@@ -1,15 +1,8 @@
...
   list(APPEND Caffe2_CPU_SRCS
     utils/string_utils.cc
-    utils/threadpool/pthreadpool.cc
-    utils/threadpool/pthreadpool_impl.cc
-    utils/threadpool/pthreadpool_new_if_impl.c
```
Comment (AshkanAliabadi) [inline caffe2/utils/CMakeLists.txt:9]: We don't need these alternate implementations of pthreadpool compiled on a mobile build.

File: caffe2/utils/CMakeLists.txt [hunk 8]
```diff
@@ -1,15 +1,8 @@
...
+    utils/threadpool/pthreadpool-cpp.cc
     utils/threadpool/ThreadPool.cc
-    utils/threadpool/ThreadPoolMobile.cc
-    utils/threadpool/ThreadPoolXNNPACK.cc
```
Comment (AshkanAliabadi) [inline caffe2/utils/CMakeLists.txt:12]: Removed ThreadPoolXNNPACK.  Not needed anymore.

File: caffe2/utils/CMakeLists.txt [hunk 9]
```diff
@@ -1,15 +1,8 @@
...
   )
   set(Caffe2_CPU_SRCS ${Caffe2_CPU_SRCS} PARENT_SCOPE)
   return()
```
Comment (AshkanAliabadi) [inline caffe2/utils/CMakeLists.txt:8]: The mobile build returns here.  The rest of this file is non-mobile only.

File: caffe2/utils/CMakeLists.txt [hunk 10]
```diff
@@ -28,23 +21,16 @@ list(APPEND Caffe2_CPU_SRCS
   utils/proto_convert.cc
   utils/proto_utils.cc
   utils/proto_wrap.cc
+  utils/threadpool/pthreadpool-cpp.cc
+  utils/threadpool/ThreadPool.cc
   utils/signal_handler.cc
   utils/smart_tensor_printer.cc
-  utils/string_utils.cc
-  utils/threadpool/ThreadPool.cc)
+  utils/string_utils.cc)
```
Comment (AshkanAliabadi) [inline caffe2/utils/CMakeLists.txt:28]: If we are on a non-mobile build (a mobile build cannot reach here) we still need pthreadpool-cpp.cc since NNPACK, QNNPACK, and XNNPACK are not mobile specific.

File: caffe2/utils/threadpool/pthreadpool-cpp.cc [hunk 11]
```diff
@@ -0,0 +1,75 @@
+#include <caffe2/utils/threadpool/pthreadpool-cpp.h>
+#include <c10/util/Exception.h>
+
```
Comment (AshkanAliabadi) [inline caffe2/utils/threadpool/pthreadpool-cpp.cc:3]: All of the pthreadpool symbols present in this file, namely pthreadpool_create, pthreadpool_destroy, pthreadpool_get_threads_count, and pthreadpool_parallelize_1d have a C2 and third party implementation.  The build system picks which one to use.  This implementation is oblivious to that.

File: caffe2/utils/threadpool/pthreadpool.cc [hunk 12]
```diff
@@ -86,7 +86,7 @@ static void compute_2d(void* context_, size_t linear_index) {
 }
 
 void pthreadpool_compute_2d(
-  struct pthreadpool* threadpool,
+  pthreadpool_t threadpool,
```
Comment (AshkanAliabadi) [inline caffe2/utils/threadpool/pthreadpool.cc:89]: This was a typo.

File: caffe2/utils/threadpool/pthreadpool.h [hunk 13]
```diff
@@ -33,48 +33,6 @@ typedef void (*pthreadpool_function_4d_tiled_t)(
...
-typedef void (*pthreadpool_task_6d_tile_2d_t)(
-    void*,
-    size_t,
-    size_t,
-    size_t,
-    size_t,
-    size_t,
-    size_t,
-    size_t,
-    size_t);
```
Comment (AshkanAliabadi) [inline caffe2/utils/threadpool/pthreadpool.h:76]: The extended pthreadpool API is no longer needed for C2's implementation, as was the case before.

File: caffe2/utils/threadpool/pthreadpool.h [hunk 14]
```diff
@@ -123,6 +81,13 @@ void pthreadpool_compute_1d(
     void* argument,
     size_t range);
 
+void pthreadpool_parallelize_1d(
+    pthreadpool_t threadpool,
+    pthreadpool_function_1d_t function,
+    void* argument,
+    size_t range,
+    uint32_t flags);
+
```
Comment (AshkanAliabadi) [inline caffe2/utils/threadpool/pthreadpool.h:90]: Since we are using this one function from the extended version, add it here.

File: cmake/Dependencies.cmake [hunk 15]
```diff
@@ -274,44 +274,59 @@ if(INTERN_BUILD_MOBILE AND INTERN_USE_EIGEN_BLAS)
 endif()
```
Comment (AshkanAliabadi) [inline cmake/Dependencies.cmake:275]: This is the most important file in this PR.  This logic is complicated and picks what implementation we want to select.

File: cmake/Dependencies.cmake [hunk 16]
```diff
@@ -274,44 +274,59 @@ if(INTERN_BUILD_MOBILE AND INTERN_USE_EIGEN_BLAS)
...
-    set(CAFFE2_THIRD_PARTY_ROOT "${PROJECT_SOURCE_DIR}/third_party")
-    set(PTHREADPOOL_SOURCE_DIR "${CAFFE2_THIRD_PARTY_ROOT}/pthreadpool" CACHE STRING "pthreadpool source directory")
-  endif()
+    if(NOT TARGET pthreadpool)
+      set(PTHREADPOOL_BUILD_TESTS OFF CACHE BOOL "")
+      set(PTHREADPOOL_BUILD_BENCHMARKS OFF CACHE BOOL "")
+      add_subdirectory(
+        "${PTHREADPOOL_SOURCE_DIR}"
+        "${CONFU_DEPENDENCIES_BINARY_DIR}/pthreadpool")
+      set_property(TARGET pthreadpool PROPERTY POSITION_INDEPENDENT_CODE ON)
```
Comment (AshkanAliabadi) [inline cmake/Dependencies.cmake:326]: PIC property now factored out from all usage below and put here.

File: cmake/Dependencies.cmake [hunk 17]
```diff
@@ -274,44 +274,59 @@ if(INTERN_BUILD_MOBILE AND INTERN_USE_EIGEN_BLAS)
...
+# BUILD_CAFFE2_OPS is on.  This will force both PyTorch and Caffe2 to use the
+# Caffe2 pthreadpool version (remember we cannot have two copies in one binary)
+# and in doing so we achieve our goal of keeping Caffe2's behavior intact if
+# at the cost of some performance loss to PyTorch.  Otherwise, if BUILD_CAFFE2_OPS
+# is not requested, we take the liberty of picking the open source version to
+# improve PyTorch's performance.  All of this admittedly error prone logic
+# can be removed and simplified if and when we decide to migrate Caffe2 to the
+# open source pthreadpool as well.
+
+if(NOT USE_SYSTEM_PTHREADPOOL)
```
Comment (AshkanAliabadi) [inline cmake/Dependencies.cmake:295]: A contributor recently added USE_SYSTEM_PTHREADPOOL to mean the one installed on the Linux distribution.  So we now have the choice between C2's implementation, the system implementation, and third party implementation!  This complicates everything.  I am doing my best to handle it here.

File: cmake/Dependencies.cmake [hunk 18]
```diff
@@ -440,7 +457,6 @@ if(USE_XNNPACK AND NOT USE_SYSTEM_XNNPACK)
   endif()
 
   if(NOT TARGET XNNPACK)
-    set(XNNPACK_CUSTOM_THREADPOOL ON CACHE BOOL "")
```
Comment (AshkanAliabadi) [inline cmake/Dependencies.cmake:443]: There is no such thing as XNNPACK_CUSTOM_THREADPOOL.  This flag is non existent.

File: cmake/TorchConfig.cmake.in [hunk 19]
```diff
@@ -69,6 +69,11 @@ if(NOT @BUILD_SHARED_LIBS@)
     list(APPEND TORCH_LIBRARIES ${XNNPACK_LIBRARY})
   endif()
 
+  if(NOT @USE_INTERNAL_PTHREADPOOL_IMPL@)
+    find_library(PTHREADPOOL_LIBRARY pthreadpool PATHS "${TORCH_INSTALL_PREFIX}/lib")
+    list(APPEND TORCH_LIBRARIES ${PTHREADPOOL_LIBRARY})
+  endif()
```
Comment (AshkanAliabadi) [inline cmake/TorchConfig.cmake.in:75]: This is for installations of PyTorch.  We may or may not end up having a dependency on the third party pthreadpool implementation.  In case we do, we need to add it to the list of libraries a client of PyTorch should link with.

File: cmake/Dependencies.cmake (hunk 17 above)
Comment (AshkanAliabadi) [reply cmake/Dependencies.cmake:295]: @cdluminate Please review.

**Comment by kimishpatel** [comment]
> That seems to have given C2's implementation the upper hand in performance, hence justifying the added maintenance complexity, until the third party version improved in parallel surpassing the efficiency of C2's implementation as I have verified in benchmarks.

Regarding this, do we understand what gives better performance for OSS pthreadpool? Is it the use of futex vs std::conditional_variable (not sure if that is implemented using futex)?

**Comment by kimishpatel** [comment]
BTW, thanks for the summary. It provides the context of the change quite well.

File: caffe2/utils/threadpool/pthreadpool-cpp.cc [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+      range,
+      0u);
+}
+
+// Forward declaration
+size_t getDefaultNumThreads();
+
+PThreadPool* pthreadpool() {
+  static std::unique_ptr<PThreadPool> threadpool =
+      std::make_unique< PThreadPool>(getDefaultNumThreads());
```
Comment (kimishpatel) [inline caffe2/utils/threadpool/pthreadpool-cpp.cc:65]: Extra space before `PThreadPool`

File: caffe2/utils/threadpool/pthreadpool-cpp.cc [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+      &context,
+      range,
+      0u);
+}
+
+// Forward declaration
+size_t getDefaultNumThreads();
+
+PThreadPool* pthreadpool() {
+  static std::unique_ptr<PThreadPool> threadpool =
```
Comment (kimishpatel) [inline caffe2/utils/threadpool/pthreadpool-cpp.cc:64]: Might be a stupid question but why wrap this in a unique_ptr?

File: caffe2/utils/threadpool/pthreadpool-cpp.cc (hunk 11 above)
Comment (kimishpatel) [reply caffe2/utils/threadpool/pthreadpool-cpp.cc:3]: Thanks for this comments Ashkan. They make it a little easier to understand. I shall adopt this :).

File: caffe2/utils/threadpool/pthreadpool-cpp.cc [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+  return pthreadpool_get_threads_count(threadpool_.get());
+}
+
+void PThreadPool::set_thread_count(const size_t thread_count) {
+  // As it stands, pthreadpool is an entirely data parallel framework with no
+  // support for task parallelism.  Hence, all functions are blocking, and no
+  // user-provided tasks can be in flight when the control is returned to the
+  // user of the API, which means re-initializing the library, without the
+  // need to wait on any pending tasks, is all one needs to do to re-adjust
+  // the thread count.
```
Comment (kimishpatel) [inline caffe2/utils/threadpool/pthreadpool-cpp.cc:21]: Thanks for the comment.

File: caffe2/utils/threadpool/pthreadpool-cpp.h [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+  PThreadPool& operator=(const PThreadPool&) = delete;
+
+  PThreadPool(PThreadPool&&) = default;
+  PThreadPool& operator=(PThreadPool&&) = default;
+
+  size_t get_thread_count() const;
+  void set_thread_count(size_t thread_count);
+
+  // Run, in parallel, function fn(task_id) over task_id in range [0, range).
+  // This function is blocking.  All input is processed by the time it returns.
```
Comment (kimishpatel) [inline caffe2/utils/threadpool/pthreadpool-cpp.h:29]: double spaces in the comment at "  All".

File: caffe2/utils/threadpool/pthreadpool.h (hunk 13 above)
Comment (kimishpatel) [reply caffe2/utils/threadpool/pthreadpool.h:76]: Because C2 was never using this API? It was mainly for xnnpack I think. Is that right?

File: caffe2/utils/threadpool/pthreadpool_impl.cc [hunk 24]
```diff
@@ -27,20 +27,21 @@ void pthreadpool_compute_1d(
           range);
 }
 
+void pthreadpool_parallelize_1d(
```
Comment (kimishpatel) [inline caffe2/utils/threadpool/pthreadpool_impl.cc:30]: Just for context, `pthreadpool_parallelize_1d` why is this added? Is there something on C2 side that uses this interface?

File: caffe2/utils/CMakeLists.txt [hunk 25]
```diff
@@ -28,23 +21,16 @@ list(APPEND Caffe2_CPU_SRCS
...
-# pthreadpool with a very similar interface. Neither NNPACK, nor this
-# thread pool supports Windows.
-if(NOT MSVC AND USE_XNNPACK)
-  add_definitions(-DUSE_INTERNAL_THREADPOOL_IMPL)
-  set(Caffe2_CPU_SRCS ${Caffe2_CPU_SRCS}
-          utils/threadpool/pthreadpool.cc
-          utils/threadpool/pthreadpool_impl.cc
-          utils/threadpool/pthreadpool_new_if_impl.c
-          utils/threadpool/ThreadPoolMobile.cc
-          utils/threadpool/ThreadPoolXNNPACK.cc
```
Comment (kimishpatel) [inline caffe2/utils/CMakeLists.txt:46]: So do we still need `caffe2::ThreadPool` in ThreadPool.cc/h?

File: cmake/Dependencies.cmake [hunk 26]
```diff
@@ -274,44 +274,59 @@ if(INTERN_BUILD_MOBILE AND INTERN_USE_EIGEN_BLAS)
...
+# and in doing so we achieve our goal of keeping Caffe2's behavior intact if
+# at the cost of some performance loss to PyTorch.  Otherwise, if BUILD_CAFFE2_OPS
+# is not requested, we take the liberty of picking the open source version to
+# improve PyTorch's performance.  All of this admittedly error prone logic
+# can be removed and simplified if and when we decide to migrate Caffe2 to the
+# open source pthreadpool as well.
+
+if(NOT USE_SYSTEM_PTHREADPOOL)
+  # Opt for custom Caffe2 implementation whenever BUILD_CAFFE2_OPS is enabled
+  if(BUILD_CAFFE2_OPS)
```
Comment (kimishpatel) [inline cmake/Dependencies.cmake:297]: I think this is really gonna matter for internal build system? I am not sure if it matters to keep this here. It is little strange now that you cannot use xnnpack if you turn this flag ON.
For internal build I suppose we can set this for very specific builds for C2. But if we are building pytorch lib then we should not use this?
Also I am wondering if this is further going to complicate other things when we dispatch to C2 ops from PyTorch as I think Linbin was trying. @linbinyu.
My apologies here. I know this has already been complicated enough so dont want to throw more troubles here, but let's make sure to bring all the stakeholders here and make sure it does not build existing things.

File: cmake/Dependencies.cmake [hunk 27]
```diff
@@ -393,19 +410,19 @@ if(USE_PYTORCH_QNNPACK)
     endif()
 
     if(NOT TARGET pytorch_qnnpack)
+      if(NOT USE_SYSTEM_PTHREADPOOL AND USE_INTERNAL_PTHREADPOOL_IMPL)
```
Comment (kimishpatel) [inline cmake/Dependencies.cmake:413]: Do we want PYTORCH_QNNPACK to have same behavior as XNNPACK? That is make it always use OSS? Does not have to be in this PR but it will require changes to our fork of QNNPACK. Maybe we can sync up and decide.

File: cmake/External/nnpack.cmake [hunk 28]
```diff
@@ -59,9 +59,12 @@ if(ANDROID OR IOS OR ${CMAKE_SYSTEM_NAME} STREQUAL "Linux" OR ${CMAKE_SYSTEM_NAM
   set(GOOGLETEST_SOURCE_DIR "${CAFFE2_THIRD_PARTY_ROOT}/googletest" CACHE STRING "Google Test source directory")
 
   if(NOT TARGET nnpack)
+    if(NOT USE_SYSTEM_PTHREADPOOL AND USE_INTERNAL_PTHREADPOOL_IMPL)
```
Comment (kimishpatel) [inline cmake/External/nnpack.cmake:62]: When is this needed? For C2's NNPACK integration? If so can we use BUILD_CAFFE2_OPS?

File: aten/src/ATen/ParallelNative.cpp [hunk 29]
```diff
@@ -87,15 +86,15 @@ void _run_with_pool(const std::function<void(int, size_t)>& fn, size_t range) {
...
+
+  pool->run(
+    // PThreadPool::run() is blocking.  A std::function [const] reference to
```
Comment (kimishpatel) [inline aten/src/ATen/ParallelNative.cpp:93]: Why not maintain the same function signature as (int, size_t) instead of constructing another wrapper lambda?

File: aten/src/ATen/ParallelNative.cpp (hunk 3 above)
Comment (kimishpatel) [reply aten/src/ATen/ParallelNative.cpp:90]: So are we ensuring that on windows platform we are building C2's implementation?

File: aten/src/ATen/ParallelNative.cpp [hunk 30]
```diff
@@ -277,8 +278,8 @@ std::shared_ptr<c10::ivalue::Future> intraop_launch_future(
   }
   return future;
 #else
-  // TODO: caffe2::ThreadPool doesn't support submitting tasks separately and
-  // running in parallel. Should fix it when this API becomes popular.
+  // TODO: caffe2::PThreadPool only provides a data-parallel API.
+  // Task parallelism is not currently supported.
```
Comment (kimishpatel) [inline aten/src/ATen/ParallelNative.cpp:282]: This is mainly used in `at::launch`. I dont think we are there yet to support it on mobile.

Comment (kimishpatel) [reply aten/src/ATen/ParallelNative.cpp:282]: And only fork/wait ops use this.

File: caffe2/utils/CMakeLists.txt [hunk 31]
```diff
@@ -1,15 +1,8 @@
...
-# to pass the actual threadpool ptr instead of nullptr.
 if(INTERN_BUILD_MOBILE AND NOT BUILD_CAFFE2_MOBILE)
-  add_definitions(-DUSE_INTERNAL_THREADPOOL_IMPL)
   list(APPEND Caffe2_CPU_SRCS
     utils/string_utils.cc
-    utils/threadpool/pthreadpool.cc
-    utils/threadpool/pthreadpool_impl.cc
-    utils/threadpool/pthreadpool_new_if_impl.c
+    utils/threadpool/pthreadpool-cpp.cc
     utils/threadpool/ThreadPool.cc
```
Comment (kimishpatel) [inline caffe2/utils/CMakeLists.txt:5]: Do we need ThreadPool.cc

**Review by kimishpatel — CHANGES_REQUESTED** [review]
This is looking good. I have left some comments.

File: aten/src/ATen/ParallelNative.cpp (hunk 3 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/ParallelNative.cpp:90]: Third party pthreadpool shouldn't build on Windows because POSIX headers such as pthreads.h and unistd.h are not available there.  I haven't tested Windows locally, but I am taking Windows CIs passing to mean that at least something is not horribly broken. :)

File: caffe2/utils/threadpool/pthreadpool.h (hunk 13 above)
Comment (AshkanAliabadi) [reply caffe2/utils/threadpool/pthreadpool.h:76]: Yes, that's correct.  This new API is only used for XNNPACK, which is now using the third-party version.

File: caffe2/utils/threadpool/pthreadpool-cpp.cc (hunk 11 above)
Comment (AshkanAliabadi) [reply caffe2/utils/threadpool/pthreadpool-cpp.cc:3]: > > That seems to have given C2's implementation the upper hand in performance, hence justifying the added maintenance complexity, until the third party version improved in parallel surpassing the efficiency of C2's implementation as I have verified in benchmarks.
> 
> Regarding this, do we understand what gives better performance for OSS pthreadpool? Is it the use of futex vs std::conditional_variable (not sure if that is implemented using futex)?

I think it is just a bug in C2's implementation but I am not certain.  That's just my guess.

File: caffe2/utils/threadpool/pthreadpool-cpp.cc (hunk 21 above)
Comment (AshkanAliabadi) [reply caffe2/utils/threadpool/pthreadpool-cpp.cc:64]: So as to not leak the memory.  Not that it really matters since this variable's lifetime is the entirety of the program.

File: caffe2/utils/CMakeLists.txt (hunk 25 above)
Comment (AshkanAliabadi) [reply caffe2/utils/CMakeLists.txt:46]: Yes, only because it's got a single function we use, namely getDefaultNumThreads().  If I had moved that function out of ThreadPool.cc and into a common separate file we wouldn't need that file anymore.

File: cmake/Dependencies.cmake (hunk 26 above)
Comment (AshkanAliabadi) [reply cmake/Dependencies.cmake:297]: Thank you for your comment Kimish.  These are very valid points.  I think this is the trickiest part of this PR.  Since we cannot have two copies of pthreadpool, we have to somehow surgically decide when to use what.  Alternatively we can just use third-party pthreadpool in CMake for all OSS use cases, regardless of whether we use C2 or PT, and only concentrate on making the C2 vs PT separation for internal usecases.  As a matter of fact that would make things mich easier.  The reason a number of builds are currently failing is because I need to modify the JNI builds to only link against third_party pthreadpools and XNNPACK, only if they are available which is no longer the case with these changes.

@dreiss What do you think?

File: cmake/Dependencies.cmake (hunk 27 above)
Comment (AshkanAliabadi) [reply cmake/Dependencies.cmake:413]: The only downside as far as I can tell is that we would loose the ability to parallelize QNNPACK and NNPACK on Windows, where third-party pthreadpool is not supported.  If we are willing to leave Windows behind for NNPACK and QNNPACK we can do that.

File: cmake/External/nnpack.cmake (hunk 28 above)
Comment (AshkanAliabadi) [reply cmake/External/nnpack.cmake:62]: NNPACK is also used in PyTorch.  We want it to use the same thread pool as the rest of the code.  If we are building for C2, then NNPACK, QNNPACK, and PyTorch QNNPACK will all use C2's implementation.  Otherwise everything will use the third-party version.  So it's all this, or all that and never a mix.

The reason I am not using BUILD_CAFFE2_OPS here is because I want to centralize the logic in a single place.  That central place decides whether to USE_INTERNAL_PTHREADPOOL_IMPL or not.  If that logic has come to the conclusion to use C2's implementation, either because BUILD_CAFFE2_OPS is ON, or we are on Windows, or whatever other condition, then that sets the USE_INTERNAL_PTHREADPOOL_IMPL flag and that way at least we wouldn't have to duplicate that error prone logic everywhere.

File: aten/src/ATen/ParallelNative.cpp (hunk 29 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/ParallelNative.cpp:93]: This code went through several iterations and maybe I would have done it differently if development had followed a different course. :)  Having said that, the main reason I left the first argument out is mainly because it seems to no longer be used at all.  Even this usage of the thread pool has no use for that argument and just hardcodes the argument to 0.  I can just as well move this 0 to PThreadPool.

File: aten/src/ATen/ParallelNative.cpp (hunk 30 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/ParallelNative.cpp:282]: Thanks.

File: caffe2/utils/CMakeLists.txt (hunk 31 above)
Comment (AshkanAliabadi) [reply caffe2/utils/CMakeLists.txt:5]: Yes, only because it's got a single function we use, namely getDefaultNumThreads(). If I had moved that function out of ThreadPool.cc and into a common separate file we wouldn't need that file anymore.

File: cmake/Dependencies.cmake (hunk 27 above)
Comment (AshkanAliabadi) [reply cmake/Dependencies.cmake:413]: Actually coming to think about it, I think NNPACK and family do not build on Windows to begin with!

File: caffe2/utils/threadpool/pthreadpool-cpp.cc (hunk 22 above)
Comment (dreiss) [reply caffe2/utils/threadpool/pthreadpool-cpp.cc:21]: What if one thread calls `set_thread_count` while another thread is running a model?

File: cmake/Dependencies.cmake (hunk 27 above)
Comment (dreiss) [reply cmake/Dependencies.cmake:413]: QNNPACK works on Windows.  I don't think we care about *NNPACK performance on Windows builds.

File: cmake/Dependencies.cmake (hunk 26 above)
Comment (dreiss) [reply cmake/Dependencies.cmake:297]: I think it's fine to move everything to the new third party pthreadpool for all OSS builds.

It looks like none of the ops that Linbin will be affected.  GenerateProposals doesn't appear to use threads, and RoIAlign only uses OpenMP.

**Review by dreiss — CHANGES_REQUESTED** [review]
We should talk about what the plan is for internal builds before landing this.

**Review by dreiss — APPROVED** [review]
Sorry, meant to approve.
